modules/javafx.graphics/src/main/java/com/sun/glass/ui/headless/HeadlessRobot.java

```diff
@@ -0,0 +1,398 @@
+        if (activeWindow == null) return;
+        HeadlessView view = (HeadlessView)activeWindow.getView();
+        int code = keyCode.getCode();
+        processSpecialKeys(code, true);
+        char[] keyval = getKeyChars(code);
+        int mods = getKeyModifiers();
+        if (view != null) {
+            view.notifyKey(KeyEvent.PRESS, code, keyval, mods);
+            if (keyval.length > 0) {
+                view.notifyKey(KeyEvent.TYPED, 0, keyval, mods);
+            }
+        }
+    }
+
+    @Override
+    public void keyRelease(KeyCode keyCode) {
+        checkWindowFocused();
+        if (activeWindow == null) return;
+        HeadlessView view = (HeadlessView)activeWindow.getView();
+        int code = keyCode.getCode();
+        processSpecialKeys(code, false);
+        int mods = getKeyModifiers();
+        char[] keyval = new char[1];
+        keyval[0] = (char) code;
+        if (view != null) {
+            view.notifyKey(KeyEvent.RELEASE, code, keyval, mods);
+        }
+    }
+
+    @Override
+    public double getMouseX() {
+        return this.mouseX;
+    }
+
+    @Override
+    public double getMouseY() {
+        return this.mouseY;
+    }
+
+    @Override
+    public void mouseMove(double x, double y) {
+        this.mouseX = x;
+        this.mouseY = y;
+        checkWindowEnterExit();
+        if (activeWindow == null) return;
+        HeadlessView view = (HeadlessView)activeWindow.getView();
+        if (view == null) return;
+        int wx = activeWindow.getX();
+        int wy = activeWindow.getY();
+        int buttonEvent = MouseEvent.BUTTON_NONE;
+        int mouseEvent = MouseEvent.MOVE;
+        if (mouseState.pressedButtons.size() > 0) {
+            MouseButton button = mouseState.pressedButtons.stream().findFirst().get();
+            buttonEvent = getGlassEventButton(button);
+            mouseEvent = MouseEvent.DRAG;
+        }
+        int modifiers = 0;
+        view.notifyMouse(mouseEvent, buttonEvent, (int)mouseX-wx, (int)mouseY-wy, (int)mouseX, (int)mouseY, modifiers, false, false);
+    }
+
+    @Override
+    public void mousePress(MouseButton... buttons) {
+        Application.checkEventThread();
+        mouseState.pressedButtons.addAll(Arrays.asList(buttons));
+        checkWindowEnterExit();
+        HeadlessView view = (HeadlessView)activeWindow.getView();
+        if (view == null) {
+            view = (HeadlessView)activeWindow.getView();
+            if (view == null) {
+                return;
+            }
+        }
+        int wx = activeWindow.getX();
+        int wy = activeWindow.getY();
+        int modifiers = getModifiers(buttons);
+        view.notifyMouse(MouseEvent.DOWN, getGlassEventButton(buttons), (int)mouseX-wx, (int)mouseY-wy, (int)mouseX, (int)mouseY, modifiers, true, true);
+        int buttonCode = getGlassEventButton(buttons);
+        view.notifyMouse(MouseEvent.DOWN, buttonCode, (int)mouseX-wx, (int)mouseY-wy, (int)mouseX, (int)mouseY, modifiers, true, true);
```

Comment:
This code (`view.notifyMouse(...`) is duplicated, and there are two notifications of mouse down while there should be just one.

The first line can be removed.